README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [proxy-handler-decorators](#proxy-handler-decorators)
  - [Functions](#functions)
    - [AllowGetTarget](#allowgettarget)
    - [DefaultToPrimitive](#defaulttoprimitive)
    - [DefaultToStringTag](#defaulttostringtag)
    - [Mapped](#mapped)
    - [TargetAsThis](#targetasthis)
    - [getTarget](#gettarget)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->


<a name="readmemd"></a>

# proxy-handler-decorators

Class decorators for proxy handlers.

[![npm package][npm-img]][npm-url]
[![Build Status][build-img]][build-url]
[![Downloads][downloads-img]][downloads-url]
[![Issues][issues-img]][issues-url]
[![Code Coverage][codecov-img]][codecov-url]
[![Commitizen Friendly][commitizen-img]][commitizen-url]
[![Semantic Release][semantic-release-img]][semantic-release-url]

[build-img]:https://github.com/Atry/proxy-handler-decorators/actions/workflows/release.yml/badge.svg
[build-url]:https://github.com/Atry/proxy-handler-decorators/actions/workflows/release.yml
[downloads-img]:https://img.shields.io/npm/dt/proxy-handler-decorators
[downloads-url]:https://www.npmtrends.com/proxy-handler-decorators
[npm-img]:https://img.shields.io/npm/v/proxy-handler-decorators
[npm-url]:https://www.npmjs.com/package/proxy-handler-decorators
[issues-img]:https://img.shields.io/github/issues/Atry/proxy-handler-decorators
[issues-url]:https://github.com/Atry/proxy-handler-decorators/issues
[codecov-img]:https://codecov.io/gh/Atry/proxy-handler-decorators/branch/main/graph/badge.svg
[codecov-url]:https://codecov.io/gh/Atry/proxy-handler-decorators
[semantic-release-img]:https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
[semantic-release-url]:https://github.com/semantic-release/semantic-release
[commitizen-img]:https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
[commitizen-url]:http://commitizen.github.io/cz-cli/

**`Example`**

A proxy to transform an array, prepending a '!' character to each element.

```typescript doctest
import {
  getTarget,
  AllowGetTarget,
  Mapped,
  DefaultToPrimitive,
  DefaultToStringTag,
  TargetAsThis,
} from 'proxy-handler-decorators';
import { DefaultProxyHandler } from 'default-proxy-handler';

let counter = 0;
function addBang(words: string[]): string[] {
  counter++;
  return words.map((word) => `${word}!`);
}

@AllowGetTarget
@Mapped<string[]>(addBang)
@DefaultToPrimitive
@DefaultToStringTag
@TargetAsThis
class MappedProxyHandler extends DefaultProxyHandler<string[]> {}
const mappedProxyHandler = new MappedProxyHandler();

const proxy = new Proxy(['hello', 'world'], mappedProxyHandler);
expect(counter).toBe(0);
expect(proxy).toEqual(['hello!', 'world!']);
expect(counter).toBeGreaterThan(0);

counter = 0;
expect(proxy).toEqual(['hello!', 'world!']);
expect(counter).toBeGreaterThan(0);

expect(getTarget(proxy)).toEqual(['hello', 'world']);
```

## Functions

### AllowGetTarget

▸ **AllowGetTarget**<`T`, `SuperClass`\>(`superClass`): `SuperClass`

The decorator to enable [getTarget](#gettarget) function.

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |
| `SuperClass` | extends (...`args`: `any`[]) => `ProxyHandler`<`T`\> |

#### Parameters

| Name | Type |
| :------ | :------ |
| `superClass` | `SuperClass` |

#### Returns

`SuperClass`

#### Defined in

[index.ts:194](https://github.com/Atry/proxy-handler-decorators/blob/8586f48/src/index.ts#L194)

___

### DefaultToPrimitive

▸ **DefaultToPrimitive**<`T`, `SuperClass`\>(`superClass`): `SuperClass`

A decorator to add a default implementation of `[Symbol.toPrimitive]` method
to the proxy.

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |
| `SuperClass` | extends (...`args`: `any`[]) => `ProxyHandler`<`T`\> |

#### Parameters

| Name | Type |
| :------ | :------ |
| `superClass` | `SuperClass` |

#### Returns

`SuperClass`

#### Defined in

[index.ts:114](https://github.com/Atry/proxy-handler-decorators/blob/8586f48/src/index.ts#L114)

___

### DefaultToStringTag

▸ **DefaultToStringTag**<`T`, `SuperClass`\>(`superClass`): `SuperClass`

A decorator to add a default implementation of `[Symbol.toStringTag]` method
to the proxy.

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |
| `SuperClass` | extends (...`args`: `any`[]) => `ProxyHandler`<`T`\> |

#### Parameters

| Name | Type |
| :------ | :------ |
| `superClass` | `SuperClass` |

#### Returns

`SuperClass`

#### Defined in

[index.ts:150](https://github.com/Atry/proxy-handler-decorators/blob/8586f48/src/index.ts#L150)

___

### Mapped

▸ **Mapped**<`T`\>(`mapper`): (`proxyHandlerConstructor`: (...`args`: `any`[]) => `ProxyHandler`<`T`\>) => `void`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `mapper` | (`from`: `T`) => `T` |

#### Returns

`fn`

▸ (`proxyHandlerConstructor`): `void`

##### Parameters

| Name | Type |
| :------ | :------ |
| `proxyHandlerConstructor` | (...`args`: `any`[]) => `ProxyHandler`<`T`\> |

##### Returns

`void`

#### Defined in

[index.ts:212](https://github.com/Atry/proxy-handler-decorators/blob/8586f48/src/index.ts#L212)

___

### TargetAsThis

▸ **TargetAsThis**<`T`, `SuperClass`\>(`superClass`): `SuperClass`

A decorator to bind `this` to the proxy target for all instance methods on
the proxy.

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `object` |
| `SuperClass` | extends (...`args`: `any`[]) => `ProxyHandler`<`T`\> |

#### Parameters

| Name | Type |
| :------ | :------ |
| `superClass` | `SuperClass` |

#### Returns

`SuperClass`

#### Defined in

[index.ts:78](https://github.com/Atry/proxy-handler-decorators/blob/8586f48/src/index.ts#L78)

___

### getTarget

▸ **getTarget**<`T`\>(`maybeProxy`): `T` \| `undefined`

Returns the proxy target if `proxy` is a proxy and its proxy handler has the
[AllowGetTarget](#allowgettarget) decorator, or `undefined` otherwise.

#### Type parameters

| Name |
| :------ |
| `T` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `maybeProxy` | `T` |

#### Returns

`T` \| `undefined`

#### Defined in

[index.ts:187](https://github.com/Atry/proxy-handler-decorators/blob/8586f48/src/index.ts#L187)


CHANGELOG.md
## [1.0.4](https://github.com/Atry/proxy-handler-decorators/compare/v1.0.3...v1.0.4) (2023-01-16)


### Dependencies and Other Build Updates

* update keywords ([4fc490d](https://github.com/Atry/proxy-handler-decorators/commit/4fc490d4bf66d80a7a17b15bf24cab66055c02f8))

## [1.0.3](https://github.com/Atry/proxy-handler-decorators/compare/v1.0.2...v1.0.3) (2023-01-16)


### Dependencies and Other Build Updates

* **deps:** remove unused doctest-ts-improved ([4a04f6e](https://github.com/Atry/proxy-handler-decorators/commit/4a04f6e78e8362c32f3c193f9c5b234ba0ef7f66))

## [1.0.2](https://github.com/Atry/proxy-handler-decorators/compare/v1.0.1...v1.0.2) (2023-01-16)

## [1.0.1](https://github.com/Atry/proxy-handler-decorators/compare/v1.0.0...v1.0.1) (2023-01-16)

## 1.0.0 (2023-01-16)


### Features

* **index:** change to bump minor version ([#7](https://github.com/Atry/proxy-handler-decorators/issues/7)) ([84cb8c4](https://github.com/Atry/proxy-handler-decorators/commit/84cb8c44d0dd5143ecd1d12b129b426283d10ef5))
* **template:** initial commit ([e37f2fb](https://github.com/Atry/proxy-handler-decorators/commit/e37f2fbb4998ac87a8c00bf032b30163d2e809d2))


### Bug Fixes

* **cve:** upgrade semantic-release to fix GHSA-rrrm-qjm4-v8hf ([d9856cd](https://github.com/Atry/proxy-handler-decorators/commit/d9856cd7ce303bad4dbac29c29ee01d97cd4b9fa))
* **deps:** cve-2021-23425 ([ecc8553](https://github.com/Atry/proxy-handler-decorators/commit/ecc8553254d6724195472dd644f2e83df16cb93c))
* **deps:** cve-2021-23440 ([8c88925](https://github.com/Atry/proxy-handler-decorators/commit/8c889253bf6a9a5209816b13398b7f7f1cdae416))
* **deps:** update caniuse db for jest deprecation warning ([29eb3cc](https://github.com/Atry/proxy-handler-decorators/commit/29eb3ccdf8e64243095f78a953e1498b5487f044))
* exclude test files from build ([#19](https://github.com/Atry/proxy-handler-decorators/issues/19)) ([00517e8](https://github.com/Atry/proxy-handler-decorators/commit/00517e8aaa1b4d0754e7f99db137c7b817672e08))
* Initial implementation ([320df5e](https://github.com/Atry/proxy-handler-decorators/commit/320df5ee280cee4fa7cb14fb51dd837babd8f3e2))
* **license:** update copyright date in license ([e83518d](https://github.com/Atry/proxy-handler-decorators/commit/e83518d3dd5f0313013bacbe8fb9c68932d54945))
* ts-doctest generated wrong module path as 'xxx/index.d' ([17f6e96](https://github.com/Atry/proxy-handler-decorators/commit/17f6e960735a8b2abadbfd9061b6864fde1af93c))


### Dependencies and Other Build Updates

* **deps-dev:** bump @types/jest to 27 ([3331b8d](https://github.com/Atry/proxy-handler-decorators/commit/3331b8d4b007d8a4bd4ea0fd33de8b37b7fdda29))
* **deps-dev:** bump semantic-release from 19.0.2 to 19.0.3 ([#15](https://github.com/Atry/proxy-handler-decorators/issues/15)) ([89b8174](https://github.com/Atry/proxy-handler-decorators/commit/89b8174ecf7d951f57f72579f1ba18e627d132f4))
* **deps:** bump json5 from 2.2.0 to 2.2.3 ([#24](https://github.com/Atry/proxy-handler-decorators/issues/24)) ([5bcadc3](https://github.com/Atry/proxy-handler-decorators/commit/5bcadc3aa3cdcf8323622a0fad3aa9f33ad9bf88))
* **deps:** bump normalize-url from 6.0.0 to 6.0.1 ([#4](https://github.com/Atry/proxy-handler-decorators/issues/4)) ([8166b9a](https://github.com/Atry/proxy-handler-decorators/commit/8166b9ace81832579218ad88382e400fc9a1bdc3))
* **deps:** bump npm from 8.4.0 to 8.12.0 ([#13](https://github.com/Atry/proxy-handler-decorators/issues/13)) ([d394432](https://github.com/Atry/proxy-handler-decorators/commit/d39443238110052106418d318bee74430d9890bc))
* **deps:** bump path-parse from 1.0.6 to 1.0.7 ([#6](https://github.com/Atry/proxy-handler-decorators/issues/6)) ([9cea74b](https://github.com/Atry/proxy-handler-decorators/commit/9cea74b30748fd73b2865812d50c2cf460e24ea8))
* **deps:** bump semver-regex from 3.1.3 to 3.1.4 ([#14](https://github.com/Atry/proxy-handler-decorators/issues/14)) ([d1c7b3e](https://github.com/Atry/proxy-handler-decorators/commit/d1c7b3e29ab425de319382a81a594d73e2e51896))
* **deps:** bump trim-newlines from 3.0.0 to 3.0.1 ([#5](https://github.com/Atry/proxy-handler-decorators/issues/5)) ([2e22195](https://github.com/Atry/proxy-handler-decorators/commit/2e22195d68efd6c5b69e44bb6be5a6c0f66e4237))
* **deps:** bump ws from 7.4.5 to 7.4.6 ([#3](https://github.com/Atry/proxy-handler-decorators/issues/3)) ([63812fc](https://github.com/Atry/proxy-handler-decorators/commit/63812fc0c676e42bd2406cf90542cf6bdf639395))
* **deps:** generate readme and doctests ([b61fbc8](https://github.com/Atry/proxy-handler-decorators/commit/b61fbc8537c8ff2610dd39c057cc996ae705b530))
* **deps:** migrate to @ryansonshine/commitizen ([72419ee](https://github.com/Atry/proxy-handler-decorators/commit/72419eefb8448e7f4f09e1b2676d3288654045bc))
* **package:** remove unused run script ([81f1055](https://github.com/Atry/proxy-handler-decorators/commit/81f10554315e79bcf8292d9ae20ffedde5c1b8f8))
* **release:** add permissions to release workflow ([f05e389](https://github.com/Atry/proxy-handler-decorators/commit/f05e389f96cf273e9df7a10af9155c40200b7b39)), closes [#10](https://github.com/Atry/proxy-handler-decorators/issues/10)
* **semantic-release:** add configuration for releasing dependabot updates ([cb94c92](https://github.com/Atry/proxy-handler-decorators/commit/cb94c92147bbfc9409c8e2fee9b3f21d3b254e99))
* **tsconfig:** add comments to tsconfig ([7ae41bb](https://github.com/Atry/proxy-handler-decorators/commit/7ae41bbbcc7841ad9fd2f9b6b1c5abee93250202))
* upgrade codecov and package-lock ([8010ef8](https://github.com/Atry/proxy-handler-decorators/commit/8010ef8200b9ce30ddc89e70dc6556d6bc341f6d)), closes [#11](https://github.com/Atry/proxy-handler-decorators/issues/11)


src/index.ts
/* eslint-disable @typescript-eslint/ban-types */
/**
 * Class decorators for proxy handlers.
 *
 * [![npm package][npm-img]][npm-url]
 * [![Build Status][build-img]][build-url]
 * [![Downloads][downloads-img]][downloads-url]
 * [![Issues][issues-img]][issues-url]
 * [![Code Coverage][codecov-img]][codecov-url]
 * [![Commitizen Friendly][commitizen-img]][commitizen-url]
 * [![Semantic Release][semantic-release-img]][semantic-release-url]
 *
 * [build-img]:https://github.com/Atry/proxy-handler-decorators/actions/workflows/release.yml/badge.svg
 * [build-url]:https://github.com/Atry/proxy-handler-decorators/actions/workflows/release.yml
 * [downloads-img]:https://img.shields.io/npm/dt/proxy-handler-decorators
 * [downloads-url]:https://www.npmtrends.com/proxy-handler-decorators
 * [npm-img]:https://img.shields.io/npm/v/proxy-handler-decorators
 * [npm-url]:https://www.npmjs.com/package/proxy-handler-decorators
 * [issues-img]:https://img.shields.io/github/issues/Atry/proxy-handler-decorators
 * [issues-url]:https://github.com/Atry/proxy-handler-decorators/issues
 * [codecov-img]:https://codecov.io/gh/Atry/proxy-handler-decorators/branch/main/graph/badge.svg
 * [codecov-url]:https://codecov.io/gh/Atry/proxy-handler-decorators
 * [semantic-release-img]:https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
 * [semantic-release-url]:https://github.com/semantic-release/semantic-release
 * [commitizen-img]:https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
 * [commitizen-url]:http://commitizen.github.io/cz-cli/
 *
 * @example
 *
 * A proxy to transform an array, prepending a '!' character to each element.
 *
 * ```typescript doctest
 * import {
 *   getTarget,
 *   AllowGetTarget,
 *   Mapped,
 *   DefaultToPrimitive,
 *   DefaultToStringTag,
 *   TargetAsThis,
 * } from 'proxy-handler-decorators';
 * import { DefaultProxyHandler } from 'default-proxy-handler';
 *
 * let counter = 0;
 * function addBang(words: string[]): string[] {
 *   counter++;
 *   return words.map((word) => `${word}!`);
 * }
 *
 * @AllowGetTarget
 * @Mapped<string[]>(addBang)
 * @DefaultToPrimitive
 * @DefaultToStringTag
 * @TargetAsThis
 * class MappedProxyHandler extends DefaultProxyHandler<string[]> {}
 * const mappedProxyHandler = new MappedProxyHandler();
 *
 * const proxy = new Proxy(['hello', 'world'], mappedProxyHandler);
 * expect(counter).toBe(0);
 * expect(proxy).toEqual(['hello!', 'world!']);
 * expect(counter).toBeGreaterThan(0);
 *
 * counter = 0;
 * expect(proxy).toEqual(['hello!', 'world!']);
 * expect(counter).toBeGreaterThan(0);
 *
 * expect(getTarget(proxy)).toEqual(['hello', 'world']);
 * ```
 *
 * @module
 */

import { DecorateAll } from 'decorate-all';

/**
 * A decorator to bind `this` to the proxy target for all instance methods on
 * the proxy.
 */
export function TargetAsThis<
  T extends object,
  SuperClass extends {
    new (...args: any[]): ProxyHandler<T>;
  }
>(superClass: SuperClass): SuperClass {
  return class TargetAsThis extends superClass {
    override get(target: T, p: string | symbol, receiver: any) {
      const property: unknown = (super.get ?? Reflect.get)(target, p, receiver);
      if (typeof property === 'function') {
        const wrappedFunction = function (
          this: any,
          ...argArray: any[]
        ): unknown {
          return Reflect.apply(
            property,
            this === receiver ? target : this,
            argArray
          );
        };
        Object.defineProperty(wrappedFunction, 'name', {
          value: property.name,
          writable: false,
        });
        return wrappedFunction;
      } else {
        return property;
      }
    }
  };
}

/**
 * A decorator to add a default implementation of `[Symbol.toPrimitive]` method
 * to the proxy.
 */
export function DefaultToPrimitive<
  T extends object,
  SuperClass extends {
    new (...args: any[]): ProxyHandler<T>;
  }
>(superClass: SuperClass): SuperClass {
  return class DefaultToPrimitive extends superClass {
    override get(target: T, p: string | symbol, receiver: any) {
      const property: unknown = (super.get ?? Reflect.get)(target, p, receiver);
      if (property === undefined) {
        switch (p) {
          case Symbol.toPrimitive:
            return (hint: 'string' | 'number' | 'default') => {
              switch (hint) {
                case 'string':
                  return String(target);
                case 'number':
                  return Number(target);
                case 'default':
                  return target.valueOf();
              }
            };
          default:
            return property;
        }
      } else {
        return property;
      }
    }
  };
}

/**
 * A decorator to add a default implementation of `[Symbol.toStringTag]` method
 * to the proxy.
 */
export function DefaultToStringTag<
  T extends object,
  SuperClass extends {
    new (...args: any[]): ProxyHandler<T>;
  }
>(superClass: SuperClass): SuperClass {
  return class DefaultToStringTag extends superClass {
    override get(target: T, p: string | symbol, receiver: any) {
      const property: unknown = (super.get ?? Reflect.get)(target, p, receiver);
      if (property === undefined) {
        switch (p) {
          case Symbol.toStringTag: {
            const prototype: undefined | Record<any, unknown> =
              Object.getPrototypeOf(target);
            const constructor: unknown = prototype?.constructor;
            if (typeof constructor === 'function') {
              return constructor.name;
            } else {
              return;
            }
          }
          default:
            return property;
        }
      } else {
        return property;
      }
    }
  };
}

const getTargetSymbol = Symbol();

/**
 * Returns the proxy target if `proxy` is a proxy and its proxy handler has the
 * {@link AllowGetTarget} decorator, or `undefined` otherwise.
 */
export function getTarget<T>(maybeProxy: T): T | undefined {
  return (maybeProxy as { [getTargetSymbol]?: T })[getTargetSymbol];
}

/**
 * The decorator to enable {@link getTarget} function.
 */
export function AllowGetTarget<
  T extends object,
  SuperClass extends {
    new (...args: any[]): ProxyHandler<T>;
  }
>(superClass: SuperClass): SuperClass {
  return class AllowGetTarget extends superClass {
    override get(target: T, p: string | symbol, receiver: any) {
      switch (p) {
        case getTargetSymbol:
          return target;
        default:
          return (super.get ?? Reflect.get)(target, p, receiver) as unknown;
      }
    }
  };
}

export function Mapped<T extends object>(
  mapper: (from: T) => T
): (proxyHandlerConstructor: {
  new (...args: any[]): ProxyHandler<T>;
}) => void {
  return DecorateAll(
    (
      target: Object,
      propertyKey: string | symbol,
      descriptor: TypedPropertyDescriptor<any>
    ) => {
      if (Object.prototype.hasOwnProperty.call(Reflect, propertyKey)) {
        const originalValue: (target: T, ...restAugments: any[]) => unknown =
          descriptor.value;
        descriptor.value = (target: T, ...restAugments: any[]) =>
          originalValue(mapper(target), ...restAugments);
      }
    },
    { deep: true }
  );
}
